leaderelection/src/main/java/io/scalecube/leaderelection/RaftStateMachine.java

```diff
@@ -0,0 +1,58 @@
+package io.scalecube.leaderelection;
+
+
+import unquietcode.tools.esm.EnumStateMachine;
+
+/**
+ * RAFT state machine dealing with state transition FOLLOWER, CANNIDATE, LEADER.
+ * 
+ * @author Ronen Nachmias
+ */
+public abstract class RaftStateMachine {
+
+  private EnumStateMachine<StateType> stateMachine;
+
+  /**
+   * RaftStateMachine Contractor.
+   */
+  public RaftStateMachine() {
+    stateMachine = new EnumStateMachine<StateType>(StateType.START);
+
+    stateMachine.addTransitions(StateType.START, StateType.FOLLOWER);
+    stateMachine.addTransitions(StateType.FOLLOWER, StateType.CANDIDATE);
+    stateMachine.addTransitions(StateType.CANDIDATE, StateType.LEADER);
+    stateMachine.addTransitions(StateType.CANDIDATE, StateType.FOLLOWER);
+    stateMachine.addTransitions(StateType.LEADER, StateType.FOLLOWER);
+
+    stateMachine.addTransitions(StateType.START, StateType.STOPPED);
+    stateMachine.addTransitions(StateType.FOLLOWER, StateType.STOPPED);
+    stateMachine.addTransitions(StateType.CANDIDATE, StateType.STOPPED);
+    stateMachine.addTransitions(StateType.LEADER, StateType.STOPPED);
+
+    stateMachine.onEntering(StateType.FOLLOWER, state -> {
+        becomeFollower();
```

Comment:
leader-election thread
it happens when message trigger observer that heartbeat or vote observer is triggered